### PR TITLE
fixed conflict resolution behavior while apply podpresets

### LINKED ISSUE #47861: PodPreset application on Pod is not consistent incase of a merge conflict
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): podpreset

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**What happened**:
If you have merge conflict while applying podpresets on a Pod, current logic modifies the PodSpec partially. 
Scenario:
Let say:
 * We have a PodPreset named "inject-db-port" with environment variable DB_PORT:6000 defined.
 * We are creating a Pod P1 with two containers C1 and C2 where C1 does not have any env vars defined while C2 has env variable DB_PORT defined with value 7000.

PodPreset "inject-db-port" gets applied on Pod P1. C1's envvars are modified to inject DB_PORT:6000 and at the time of injecting DB_PORT:6000 in C2, conflict is detected and this pod-preset is ignored, but, by that time C1's envvars are already updated. So Pod P1 is partially modified in case of conflict.

**What you expected to happen**:
While applying PodPresets on a Pod, in case of merge conflict, Pod should not be modified at all. Pod should be modified only in case of successful application of PodPreset.

**How to reproduce it** (as minimally and precisely as possible):
Scenario described is good enough to reproduce this issue.

**Anything else we need to know**:
Root cause of the problem is that we have coupled the logic of merging and detecting conflict. The logic needs to be split in two phases where in first phase, we detect just the merge conflict and if there is a conflict, we reject podpreset application completely and in case of successful merge, we continue with second phase of applying (merging) Podpreset information on Pod.

### PR DESCRIPTION
**Pull request by droot**
**What this PR does / why we need it**:
This fixes the PodPreset application behavior in case of conflicts occur during the merging of Pod's information with PodPreset's. More details are in issue #47861 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes #47861

**Special notes for your reviewer**:
We are splitting the PodPreset application logic in two phases. In first phase, we try to detect the conflicts in information merging without modifying the Pod at all. If conflict occurs, then we reject the PodPresets injection. Incase of no conflicts, we apply the PodPresets and merge the information.

**Release note**:

```release-note
PodPreset is not injected if conflict occurs while applying PodPresets to a Pod.
```

### CODE REVIEW

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -27,6 +27,8 @@ ALLOW_SECURITY_CONTEXT=${ALLOW_SECURITY_CONTEXT:-""}
 PSP_ADMISSION=${PSP_ADMISSION:-""}
 NODE_ADMISSION=${NODE_ADMISSION:-""}
 RUNTIME_CONFIG=${RUNTIME_CONFIG:-""}
+# RUNTIME_CONFIG=${RUNTIME_CONFIG:-"settingsio/v1alpha1=true,settings.k8s.io/v1alpha1/podpreset=true"}
+RUNTIME_CONFIG=${RUNTIME_CONFIG:-"api/all=true"}
```
Comment (cblecker) [inline hack/local-up-cluster.sh]: What's your intended goal for "RUNTIME_CONFIG"? You're attempting to declare it on both lines 29 and 31.

Comment (droot) [inline hack/local-up-cluster.sh]: This file got checked in by mistake. I have cleaned it up and enabled "api/all" by default.  PTAL.

Comment (cblecker) [inline hack/local-up-cluster.sh]: @droot I don't see any changes to this file.

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -26,7 +26,7 @@ ALLOW_PRIVILEGED=${ALLOW_PRIVILEGED:-""}
 ALLOW_SECURITY_CONTEXT=${ALLOW_SECURITY_CONTEXT:-""}
 PSP_ADMISSION=${PSP_ADMISSION:-""}
 NODE_ADMISSION=${NODE_ADMISSION:-""}
-RUNTIME_CONFIG=${RUNTIME_CONFIG:-""}
+RUNTIME_CONFIG=${RUNTIME_CONFIG:-"api/all=true"}
```
Comment (pmorie) [inline hack/local-up-cluster.sh]: Are these debugging changes?

Comment (droot) [inline hack/local-up-cluster.sh]: Took it out. It is not needed. I thought it was needed to enable "settings..../v1alpha1" API endpoint, but I was wrong, so took it out. PTAL.

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -215,7 +215,7 @@ RKT_PATH=${RKT_PATH:-""}
 RKT_STAGE1_IMAGE=${RKT_STAGE1_IMAGE:-""}
 CHAOS_CHANCE=${CHAOS_CHANCE:-0.0}
 CPU_CFS_QUOTA=${CPU_CFS_QUOTA:-true}
-ENABLE_HOSTPATH_PROVISIONER=${ENABLE_HOSTPATH_PROVISIONER:-"false"}
+ENABLE_HOSTPATH_PROVISIONER=${ENABLE_HOSTPATH_PROVISIONER:-"true"}
```
Comment (pmorie) [inline hack/local-up-cluster.sh]: This is also a debugging change, right? Def out of scpe for this PR

File: plugin/pkg/admission/podpreset/admission.go [hunk 4]
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
...
+			errs = append(errs, err)
 		}
+	}
+	return utilerrors.NewAggregate(errs)
+}
 
-		pod.ObjectMeta.Annotations[fmt.Sprintf("%s/podpreset-%s", annotationPrefix, pip.GetName())] = pip.GetResourceVersion()
+func safeToApplyPodPresetsOnContainer(ctr *api.Container, podPresets []*settings.PodPreset) error {
+	var errs []error
+	if _, err := MergeEnv(ctr.Env, podPresets); err != nil {
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: Seems like mergeEnv is probably unintentionally exported; we should make it private unless it's used elsewhere

File: plugin/pkg/admission/podpreset/admission.go [hunk 5]
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
...
 
-	return nil
+	return utilerrors.NewAggregate(errs)
 }
 
-func mergeEnv(pip *settings.PodPreset, original []api.EnvVar) ([]api.EnvVar, error) {
-	// if there were no original envvar just return the pip envvar
-	if original == nil {
-		return pip.Spec.Env, nil
+func MergeEnv(envVars []api.EnvVar, podPresets []*settings.PodPreset) ([]api.EnvVar, error) {
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: Hrm, maybe it is intentional?

File: plugin/pkg/admission/podpreset/admission.go [hunk 6]
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
...
 
-func mergeEnvFrom(pip *settings.PodPreset, original []api.EnvFromSource) ([]api.EnvFromSource, error) {
-	// if there were no original envfrom just return the pip envfrom
-	if original == nil {
-		return pip.Spec.EnvFrom, nil
+func MergeEnvFrom(envSources []api.EnvFromSource, podPresets []*settings.PodPreset) ([]api.EnvFromSource, error) {
+
+	var mergedEnvFrom []api.EnvFromSource
+
+	mergedEnvFrom = append(mergedEnvFrom, envSources...)
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: Is there a reason you chose a different form of allocate and copy here vs mergeenv?

File: plugin/pkg/admission/podpreset/admission.go [hunk 7]
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
...
+			if !ok {
+				// if we don't already have it append it and continue
+				origVolumeMounts[v.Name] = v
+				mergedVolumeMounts = append(mergedVolumeMounts, v)
+				continue
+			}
+
+			// make sure they are identical or throw an error
+			if !reflect.DeepEqual(found, v) {
+				errs = append(errs, fmt.Errorf("merging volume mounts for %s has a conflict on %s: \n%#v\ndoes not match\n%#v\n in container", pp.GetName(), v.Name, v, found))
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: It might be helpful if this identified the 'for' as a podpreset

File: plugin/pkg/admission/podpreset/admission.go [hunk 8]
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
...
+		for _, v := range pp.Spec.VolumeMounts {
+			found, ok := origVolumeMounts[v.Name]
+			if !ok {
+				// if we don't already have it append it and continue
+				origVolumeMounts[v.Name] = v
+				mergedVolumeMounts = append(mergedVolumeMounts, v)
+				continue
+			}
+
+			// make sure they are identical or throw an error
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: I'm wondering if identical but duplicated mounts is an error - use case?

File: plugin/pkg/admission/podpreset/admission.go [hunk 9]
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
...
+			}
 		}
 	}
 
-	// key by mount path
-	orig = map[string]interface{}{}
-	for _, v := range original {
-		orig[v.MountPath] = v
-	}
+	for _, pp := range podPresets {
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: It seems like we can do all of this in a single nested loop, since this is orthogonal to the above loop but covers the same space

File: plugin/pkg/admission/podpreset/admission.go [hunk 10]
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
...
+		for _, v := range pp.Spec.Volumes {
+			found, ok := origVolumes[v.Name]
+			if !ok {
+				// if we don't already have it append it and continue
+				origVolumes[v.Name] = v
+				mergedVolumes = append(mergedVolumes, v)
+				continue
+			}
+
+			// make sure they are identical or throw an error
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: Also not sure about duplicate but equal volumes.

File: plugin/pkg/admission/podpreset/admission.go [hunk 11]
```diff
@@ -334,3 +341,35 @@ func (c *podPresetPlugin) addEvent(pod *api.Pod, pip *settings.PodPreset, messag
...
+	if pod.ObjectMeta.Annotations == nil {
+		pod.ObjectMeta.Annotations = map[string]string{}
+	}
+
+	for _, pp := range podPresets {
+		pod.ObjectMeta.Annotations[fmt.Sprintf("%s/podpreset-%s", annotationPrefix, pp.GetName())] = pp.GetResourceVersion()
+	}
+}
+
+func applyPodPresetsOnContainer(ctr *api.Container, podPresets []*settings.PodPreset) {
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: I wiuld add a comment explaining that it shiuld be safe to ignore these errors

File: plugin/pkg/admission/podpreset/admission.go (hunk 6 above)
Comment (droot) [inline plugin/pkg/admission/podpreset/admission.go]: I followed the original code and didn't change it. I am trying to dig more in to why EnvFrom are treated differently. Quick look suggests, they should also be handled in the similar ways as Env, but I need to confirm that.

**Review by droot — COMMENTED** [review]
Thanks @pmorie for the review. Have addressed all comments except one related to handling merge of EnvFrom. Will be investigating more about it today. PTAL.

**Comment by droot** [comment]
@pmorie addressed all the comments. Also updated the comment about EnvFrom. PTAL. (Also, I am not able to see review comment threads, thats very strange).

**Comment by pmorie** [comment]
@k8s-bot ok to test

File: hack/local-up-cluster.sh [hunk 12]
```diff
@@ -406,8 +406,7 @@ function start_apiserver {
     fi
 
     # Admission Controllers to invoke prior to persisting objects in cluster
-    ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount${security_admission},ResourceQuota,DefaultStorageClass,DefaultTolerationSeconds
-
+    ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,PodPreset,ServiceAccount${security_admission},ResourceQuota,DefaultStorageClass,DefaultTolerationSeconds
```
Comment (pmorie) [inline hack/local-up-cluster.sh]: If we were to do it again, I would say to make this change in another PR.

File: plugin/pkg/admission/podpreset/admission.go [hunk 13]
```diff
@@ -103,216 +104,217 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
...
+	applyPodPresetsOnPod(pod, matchingPPs)
 
-		// merge in policy for EnvFrom
-		if pip.Spec.EnvFrom != nil {
-			for i, ctr := range pod.Spec.Containers {
-				r, err := mergeEnvFrom(pip, ctr.EnvFrom)
-				if err != nil {
-					// add event to pod
-					c.addEvent(pod, pip, err.Error())
+	glog.Infof("applied podpresets successfully on Pod: %s ", pod.GetGenerateName())
```
Comment (pmorie) [inline plugin/pkg/admission/podpreset/admission.go]: Is this always going to yield the right name?

**Review by pmorie — COMMENTED** [review]
Just a couple nits; I also think godoc for each method would be fantastic to have.

File: hack/local-up-cluster.sh (hunk 12 above)
Comment (droot) [inline hack/local-up-cluster.sh]: Makes sense. Took it out. Will make a separate PR for this.

File: plugin/pkg/admission/podpreset/admission.go (hunk 13 above)
Comment (droot) [inline plugin/pkg/admission/podpreset/admission.go]: Changed it to pod.Name. Also improved the logging a bit by making it more informative (added applied podpreset names as well)

**Comment by droot** [comment]
@pmorie addressed the comments. Please take a look.

File: plugin/pkg/admission/podpreset/admission.go (hunk 13 above)
Comment (droot) [inline plugin/pkg/admission/podpreset/admission.go]: pod.Name or pod.GetName() is empty, so kept pod.GenerateName(). I guess each pod instance will get a generated name.

**Comment by pmorie** [comment]
@k8s-bot test this

**Comment by droot** [comment]
rebased with the master, so will need LGTM.
